src/wms/src/main/java/org/geoserver/wms/legendgraphic/JSONLegendGraphicBuilder.java

```diff
@@ -0,0 +1,843 @@
+                Filter filter = rule.getFilter();
+                if (filter != null) {
+                    jRule.element(FILTER, "[" + CQL.toCQL(filter) + "]");
+                }
+                if (rule.isElseFilter()) {
+                    jRule.element(ELSE_FILTER, "true");
+                }
+                JSONArray jSymbolizers = new JSONArray();
+                if (layer != null) {
+                    feature = getSampleFeatureForRule(layer, null, rule);
+                } else {
+                    feature = null;
+                }
+                List<Symbolizer> symbolizers2 = rule.symbolizers();
+                for (Symbolizer symbolizer : symbolizers2) {
+                    JSONObject jSymb = new JSONObject();
+                    JSONObject symb = processSymbolizer(symbolizer);
+                    jSymb.element(symbolizerNames.get(symbolizer.getClass()), symb);
+                    jSymbolizers.add(jSymb);
+                }
+                org.opengis.style.GraphicLegend l = rule.getLegend();
+                if (l != null) {
+                    for (GraphicalSymbol g : l.graphicalSymbols()) {
+                        JSONObject jGraphicSymb = processGraphicalSymbol(g);
+                        String href =
+                                IconPropertyExtractor.extractProperties(
+                                                gt2Style, (SimpleFeature) feature)
+                                        .href(
+                                                request.getBaseUrl(),
+                                                legend.getLayer(),
+                                                legend.getStyleName());
+
+                        jGraphicSymb.element("url", href);
+                        jRule.element(LEGEND_GRAPHIC, jGraphicSymb);
+                    }
+                }
+                if (!jSymbolizers.isEmpty()) {
+                    jRule.element(SYMBOLIZERS, jSymbolizers);
+                }
+                jRules.add(jRule);
+            }
+            if (!jRules.isEmpty()) {
+                JSONArray legends = new JSONArray();
+                if (response.containsKey(LEGEND)) {
+                    legends = response.getJSONArray(LEGEND);
+                }
+                JSONObject jLayer = new JSONObject();
+
+                jLayer.element(LAYER_NAME, layerName);
+                jLayer = getLayerTitle(jLayer, legend);
+                jLayer.element(RULES, jRules);
+                legends.add(jLayer);
+                response.element(LEGEND, legends);
+            }
+        }
+        /* }*/
+
+        return response;
+    }
+
+    private JSONObject getLayerTitle(JSONObject in, LegendRequest legend) {
+        String title = legend.getTitle();
+        return in.element(TITLE, title);
+    }
+
+    /**
+     * @param symbolizer
+     * @return
+     */
+    private String printExpression(Expression exp, Class<?> type) {
+
+        Object value = exp.evaluate(feature, type);
```

Comment:
This method seems fragile, an expression could return a static value even if attribute dependent, e.g., something containing a null check with an if function for example, or worse, could throw a NPE because the attribute needed is not available.

Better to use the ``FilterAttributeExtractor`` and see if there are attribute references, and if so, go and encode as CQL instead (not with toString thought, that's unreliable and won't encode properly all cases, use ``ECQL.toCQL``instead)